File: NinePSharp/messages/Tattach.cs
using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

namespace NinePSharp.Messages;

// size[4] Tattach tag[2] fid[4] afid[4] uname[s] aname[s]
public readonly struct Tattach : ISerializable
{
    public uint Size {get;}
    public MessageTypes Type => MessageTypes.Tattach;
    public ushort Tag {get;}

    public uint Fid {get;}
    public uint Afid {get;}
    public string Uname {get;}
    public string Aname {get;}
    public uint? NUname {get;} // 9P2000.u extension

    public Tattach(ushort tag, uint fid, uint afid, string uname, string aname)
    {
        Tag = tag;
        Fid = fid;
        Afid = afid;
        Uname = uname;
        Aname = aname;
        NUname = null;
        Size = (uint)(NinePConstants.HeaderSize + 4 + 4 + 2 + System.Text.Encoding.UTF8.GetByteCount(uname) + 2 + System.Text.Encoding.UTF8.GetByteCount(aname));
    }

    public Tattach(ReadOnlySpan<byte> data, bool is9u = false)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));

        int offset = NinePConstants.HeaderSize;
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
        offset += 4;
        
        Afid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
        offset += 4;

        Uname = data.ReadString(ref offset);
        Aname = data.ReadString(ref offset);

        if (is9u && offset + 4 <= data.Length)
        {
            NUname = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4));
            offset += 4;
        }
        else
        {
            NUname = null;
        }
    }

    public void WriteTo(Span<byte> data, bool is9u = false)
    {
        data.WriteHeaders(Size, Tag, Type);
        int offset = NinePConstants.HeaderSize;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Fid);
        offset += 4;

        BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), Afid);
        offset += 4;

        data.WriteString(Uname, ref offset);
        data.WriteString(Aname, ref offset);

        if (is9u && NUname.HasValue)
        {
            BinaryPrimitives.WriteUInt32LittleEndian(data.Slice(offset, 4), NUname.Value);
            offset += 4;
        }
    }

    void ISerializable.WriteTo(Span<byte> data) => WriteTo(data, NUname.HasValue);
}

File: NinePSharp.Messages.FSharp/messages/Tattach.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tattach =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Afid: uint32
      Uname: string
      Aname: string }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tattach
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tattach span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Afid); offset <- offset + 4
            writeString this.Uname span &offset
            writeString this.Aname span &offset

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let afid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let uname = readString data &offset
        let aname = readString data &offset
        { Size = size; Tag = tag; Fid = fid; Afid = afid; Uname = uname; Aname = aname }

